src/api/code-challenge/content-types/code-challenge/lifecycles.js

```diff
@@ -1,43 +1,131 @@
-  return false;
-}
+const executeTests = async (
+  metaCaseCode,
+  internalTests,
+  metaLabel,
+  challengeLabel,
+  metaTestId,
+  evalResultShouldBe
+) => {
+  return pMap(
+    internalTests?.filter(removeEmpty) || [],
+    async (internalTest) => {
+      const {
+        label: internalTestLabel,
+        internalTestCode: internalTestCode,
+        id: internalTestId,
+      } = internalTest;
+
+      // console.log("<internalTest pMap> ", internalTest);
+      // console.log("newTest name: ", newTest);
+
+      const testEvaluatorResults = await runTestEvaluator({
+        metaCaseCode,
+        internalTestCode,
+        metaLabel,
+        challengeLabel,
+        metaTestId,
+        internalTest,
+        evalResultShouldBe,
+      });
+      return {
+        short: {
+          description: null,
+          ...pick(testEvaluatorResults, ["description"]),
+        },
+        long: {
+          challengeLabel,
+          metaTestId,
+          internalTestId,
+          internalTestLabel,
+          userPassed: null,
+          evalResultType: null,
+          ...pick(testEvaluatorResults, [
+            "evalResultType",
+            "userPassed",
+            "evalError",
+          ]),
+        },
+      };
+    }
+  );
+};
 
 async function getInternalTests(eventTests) {
   const internalTests = await strapi.db
     .query("challenge.code-challenge-test")
     .findMany({
-      select: ["id", "internalTest"],
+      select: ["id", "internalTest", "label"],
       where: {
         id: eventTests.map(({ id: testId }) => testId),
       },
     });
-  return internalTests;
+
+  // Originally, the content of internalTest === internalTestCode in strapi.
+  // I renamed internalTest to internalTestCode for clarity in lifecycles.js & test-evaluator.ts
+  // This renamed variable makes: (internalTestCode vs internalTestLabel vs internalTestId) the children of internalTestPackage
+
+  return internalTests
+    .map((internalTestPackage) => {
+      internalTestPackage.internalTestCode = internalTestPackage.internalTest;
+      delete internalTestPackage.internalTest;
+      return internalTestPackage;
+    })
+    .sort(compareIds);
 }
 
 async function getMetaTests(eventMetaTests) {
   const metaTests = await strapi.db.query("challenge.meta-test").findMany({
-    select: ["id", "caseCode", "passes"],
+    select: ["id", "caseCode", "passes", "label"],
     where: {
       id: eventMetaTests.map(({ id: metaId }) => metaId),
     },
   });
-  return metaTests;
+  return metaTests.sort(compareIds);
 }
 
 const beforeCreateOrUpdate = async (event) => {
-  runMetaTests(event.params.data.tests, event.params.data.MetaTest);
+  // const internalLabel = await getInternalLabel(event);
+  // event.params.data.internalLabel = internalLabel;
+  conductMetaTestsOnTests(
+    event.params.data.tests,
+    event.params.data.MetaTest,
+    event.params.data.internalLabel
```

Comment:
a general best practice is to, after having more than 2 arguments in a function invocation, convert the parameters to just one: an object with properties, that way the parameters are named. Let me know if you don't understand why this is a thing

Reply:
Ok it is changed to: 

```
const beforeCreateOrUpdate = async (event) => {
  const eventParams = {
    eventTests: event.params.data.tests,
    eventMetaTests: event.params.data.MetaTest,
    challengeLabel: event.params.data.internalLabel,
  };
  runMetaTests(eventParams);
};
```
```
const runMetaTests = async (eventParams) => {
  const { eventTests, eventMetaTests, challengeLabel } = eventParams;
```